[13804] Improve instances examples (#331)
eProsima/Fast-DDS-docs: docs/fastdds/dds_layer/topic/instances.rst

.. literalinclude:: /../code/DDSCodeTester.cpp
    :language: c++
    :dedent: 12
    :start-after: //UNREGISTER-INSTANCE
    :end-before: //!

Finally, when the flight lands the instance is disposed.
This means in this specific example that as far as the DataWriter knows, the instance no longer exists and should be
considered not alive.
With this operation, the DataWriter conveys this information to the matched DataReaders.

.. literalinclude:: /../code/DDSCodeTester.cpp
    :language: c++
    :dedent: 12
    :start-after: //DISPOSE-INSTANCE
    :end-before: //!

From the management center point of view, the samples are read using the same DataReader subscribed to the Topic where
the instances are being published.
However, |SampleInfo::valid_data-api| must be checked to ensure that the sample received contains a data sample.
Otherwise, a change of the instance state is being notified.
:ref:`dds_layer_topic_instance_lifecycle` contains a diagram showing the instance statechart.

.. literalinclude:: /../code/DDSCodeTester.cpp
    :language: c++
    :dedent: 12
    :start-after: //READING-INSTANCE
    :end-before: //!

Relational databases
^^^^^^^^^^^^^^^^^^^^

> Consider now that the air traffic management center wants to keep a database with the flights being tracked.

Reviewer: In general, I'd elaborate a bit more in how each of the operations (register instance, write, unregister, and dispose) would mean in each of the two DB cases

Using DDS instances, maintaining a relational database is almost direct.
The instance key, unique identifier, is analogous to the primary key of the database.
Thus, the management center can keep the latest update for each instance in a table like the one below:

.. list-table::
    :header-rows: 1
    :align: left

    * - Primary key
      - Data
    * - ``first_flight_handle``
      - Latest location for the first flight
    * - ``second_flight_handle``

>       - Latest location for the second flight

Reviewer: Please, consider the following under a `plantuml` directive

```plantuml
@startmindmap
**:**Last position of each flight table**
----
|= Instance handle [PK] |= Data |
| 1 | Position1 |
| 2 | Position2 |
| 3 | Position3 |
| 4 | Position4 |
| 5 | Position5 |;
@endmindmap
```

Author: I have tried integrating `plantuml` mind maps with the sphinx module to no avail. I will proceed using `rst` list-tables.

In this case, every time a new sample is received, the corresponding instance entry in the database will be updated with
the latest known location.
However, the historical data of the flight track would not be easily saved into a relational database.

> If historical data is of interest, a time series database should be considered instead of a relational database.

Reviewer: I'd say relational DB can be used for this as well. Consider something as:

```plantuml
@startmindmap
**:**Topics table**
----
|= Topic name [PK] |= Instance Hnadler |
| A | 1 |
| A | 2 |
| B | 3 |
| B | 4 |
| B | 5 |;

**:**Topic A instances table**
----
|= Sequence number [PK] |= Instance handler [FK] |= Data |
| 1 | 1 | Position1 |
| 2 | 1 | Position2 |
| 3 | 2 | Position3 |
| 4 | 2 | Position4 |
| 5 | 2 | Position5 |
| 6 | 3 | Position6 |
| 7 | 3 | Position7 |;

**:**Topic B instances table**
----
|= Sequence number [PK] |= Instance handler [FK] |= Data |
| 1 | 1 | Position1 |
| 2 | 1 | Position2 |
| 3 | 1 | Position3 |
| 4 | 2 | Position4 |
| 5 | 2 | Position5 |;
@endmindmap
```

:ref:`dds_layer_topic_instance_lifecycle` contains a diagram showing the instance statechart.

.. literalinclude:: /../code/DDSCodeTester.cpp
    :language: c++
    :dedent: 12
    :start-after: //READING-INSTANCE
    :end-before: //!

Relational databases
^^^^^^^^^^^^^^^^^^^^

Consider now that the air traffic management center wants to keep a database with the flights being tracked.
Using DDS instances, maintaining a relational database is almost direct.
The instance key (unique identifier of the instance) is analogous to the primary key of the database.
Thus, the airspace traffic management center can keep the latest update for each instance in a table like the one below:

.. list-table::
    :header-rows: 1
    :align: left

    * - Instance handle [PK]
      - Data
    * - 1
      - Position1
    * - 2
      - Position2
    * - 3
      - Position3
    * - 4
      - Position4
    * - 5
      - Position5

In this case, every time a new sample is received, the corresponding instance entry in the database will be updated with
the latest known location.
Disposing the instance may translate in erasing the corresponding data from the database.

> In this scenario, registering and unregistering the instances do not reflect in the database.

Reviewer: We could add a column with the `instance_state`. That way, register and unregister may mean something

Author: Instead of adding the column to the table I have added a clarification in the text.

In the scenario being considered, the sample timestamp could be used, besides the instance handle, as primary key to
be able to access the historical tracking data of an specific flight.

.. list-table::
    :header-rows: 1
    :align: left

    * - Instance handle [PK]
      - Source Timestamp [PK]
      - Data
    * - 1
      - 1
      - Position1
    * - 2
      - 1
      - Position2
    * - 1
      - 2
      - Position3
    * - 1
      - 3
      - Position4
    * - 2
      - 2
      - Position5

In this case, looking for a specific instance handle would return the flight tracking information:

.. list-table::
    :header-rows: 1
    :align: left

    * - Instance handle [Fixed]
      - Source Timestamp
      - Data
    * - 1
      - 1
      - Position1
    * - 1
      - 2
      - Position3
    * - 1
      - 3
      - Position4

Whereas looking for a specific timestamp would allow to have a picture of the different flight locations at a specific
time:

.. list-table::
    :header-rows: 1
    :align: left

    * - Instance handle
      - Source Timestamp [Fixed]
      - Data
    * - 1
      - 2
      - Position3
    * - 2
      - 2

>       - Position5

Reviewer: I'd add a final sentence stating that dispose/unregister may mean the same as in te previous DB example

Author: I do not think is needed. It is not properly another example but another case of the same example. So the information given previously is still valid in this second scenario.

In this case, every time a new sample is received, the corresponding instance entry in the database will be updated with
the latest known location.
Disposing the instance may translate in erasing the corresponding data from the database.

> In this scenario, registering and unregistering the instances do not reflect in the database, although if the

Reviewer: ```suggestion
In this scenario, registering and unregistering the instances does not reflect in the database, although if the
```

> |SampleInfo::instance_state-api| and |SampleInfo::view_state-api| are also saved into the database the instance

Reviewer: ```suggestion
|SampleInfo::instance_state-api| and |SampleInfo::view_state-api| are also persisted the instance, then the
```
